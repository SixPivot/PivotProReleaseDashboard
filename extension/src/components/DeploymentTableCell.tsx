import React from 'react'
import { SimpleTableCell } from 'azure-devops-ui/Table'
import { Link } from 'azure-devops-ui/Link'
import { Status, StatusSize } from 'azure-devops-ui/Status'
import { AgoFormat } from 'azure-devops-ui/Utilities/Date'
import { getStatusIndicatorData } from '../utilities'
import { IDashboardEnvironmentColumn, IPipelineInstance } from '../types'
import { SafeAgo } from './SafeAgo'
import { Icon, IconSize } from 'azure-devops-ui/Icon'

interface BuildNameCellProps {
    buildName: string
    uri: string
    approvalName?: string
}

const BuildNameCell: React.FC<BuildNameCellProps> = ({ buildName, uri }) => {
    return (
        <Link className="bolt-table-inline-link bolt-table-link no-underline-link" target="_top" href={uri}>
            {buildName}
        </Link>
    )
}

interface DeploymentTableCellProps {
    columnIndex: number
    tableColumn: IDashboardEnvironmentColumn
    tableItem: IPipelineInstance
    buildName?: string
    approvalName?: string
}

export const DeploymentTableCell: React.FC<DeploymentTableCellProps> = ({
    columnIndex,
    tableColumn,
    tableItem,
    buildName,
    approvalName,
}) => {
    if (tableColumn.id === 'name') {
        return (
            <SimpleTableCell
                columnIndex={columnIndex}
                tableColumn={tableColumn}
                contentClassName="fontWeightSemiBold font-weight-semibold fontSizeM font-size-m bolt-table-cell-content-with-inline-link"
            >
                <Link className="bolt-table-inline-link bolt-table-link no-underline-link" target="_top" href={tableItem.uri}>
                    {tableItem.name}
                </Link>
            </SimpleTableCell>
        )
    }
    const env = tableItem.environments[tableColumn.id]
    return (
        <SimpleTableCell
            columnIndex={columnIndex}
            tableColumn={tableColumn}
            contentClassName="fontWeightSemiBold font-weight-semibold fontSizeM font-size-m bolt-table-cell-content-with-inline-link"
        >
            {env ? (
                <div className="flex-row flex-start">
                    <Status
                        {...getStatusIndicatorData(env.result).statusProps}
                        className="icon-large-margin status-icon"
                        size={StatusSize.m}
                    />
                    <div className="flex-column wrap-text">
                        <BuildNameCell buildName={buildName || env.value} uri={env.uri} />
                        <div className="finish-date">{env.finishTime && <SafeAgo date={env.finishTime} format={AgoFormat.Extended} />}</div>
                    </div>
                    {approvalName && (
                        <Icon
                            iconName="ReceiptCheck"
                            size={IconSize.medium}
                            style={{ color: 'blue' }}
                            tooltipProps={{ text: `Approved by: ${approvalName}` }}
                        />
                    )}
                </div>
            ) : (
                <div className="no-data">-</div>
            )}
        </SimpleTableCell>
    )
}
