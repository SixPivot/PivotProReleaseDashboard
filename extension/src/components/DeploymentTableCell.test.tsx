// At the top of your test file
const OriginalDateTimeFormat = Intl.DateTimeFormat;

jest.spyOn(Intl, 'DateTimeFormat').mockImplementation((...args) => {
  return new OriginalDateTimeFormat(['en-US'], { ...args[1], timeZone: 'UTC' });
});

import * as React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { DeploymentTableCell } from './DeploymentTableCell'
import { IDashboardEnvironmentColumn, IPipelineInstance } from '../types'

const TaskResult = { Succeeded: 2 }

const baseColumn: IDashboardEnvironmentColumn = {
    id: 'env1',
    name: 'Environment 1',
    minWidth: 100,
    maxWidth: 200,
    width: 150,
    renderCell: () => <></>,
}

const basePipeline: IPipelineInstance = {
    key: 'pipeline-1',
    name: 'Pipeline 1',
    uri: 'https://dev.azure.com/org/project/_build/results?buildId=1',
    environments: {
        env1: {
            value: 'Build 123',
            buildId: 1,
            uri: 'https://dev.azure.com/org/project/_build/results?buildId=1',
            result: TaskResult.Succeeded,
            finishTime: new Date('2025-06-29T12:00:00Z'),
            environmentId: 1,
            stageName: 'Stage 1',
        },
    },
}

describe('DeploymentTableCell', () => {
    it('renders the pipeline name cell', () => {
        render(<DeploymentTableCell columnIndex={0} tableColumn={{ ...baseColumn, id: 'name' }} tableItem={basePipeline} />)
        expect(screen.getByRole('link', { name: /Pipeline 1/i })).toBeInTheDocument()
    })

    it('renders the build name and finish date', () => {
        const { container } = render(
            <DeploymentTableCell columnIndex={1} tableColumn={baseColumn} tableItem={basePipeline} buildName="Build 123" />
        )

        expect(container).toMatchSnapshot()
    })

    it('renders the approval icon if approvalName is provided', async () => {
        render(
            <DeploymentTableCell
                columnIndex={1}
                tableColumn={baseColumn}
                tableItem={basePipeline}
                buildName="Build 123"
                approvalName="Jane Doe"
            />
        )
        // Find the icon by class or SVG
        const icon = document.querySelector('svg, .ms-Icon--ReceiptCheck')
        expect(icon).toBeInTheDocument()
    })

    it('renders a dash if no environment data', () => {
        const pipelineNoEnv: IPipelineInstance = {
            ...basePipeline,
            environments: {},
        }
        render(<DeploymentTableCell columnIndex={1} tableColumn={baseColumn} tableItem={pipelineNoEnv} />)
        expect(screen.getByText('-')).toBeInTheDocument()
    })
})
