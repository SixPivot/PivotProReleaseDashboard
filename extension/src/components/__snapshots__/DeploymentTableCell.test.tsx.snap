// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentTableCell renders the build name and finish date 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          aria-colindex="2"
          class="bolt-table-cell bolt-list-cell"
          data-column-index="1"
          role="gridcell"
        >
          <div
            class="fontWeightSemiBold font-weight-semibold fontSizeM font-size-m bolt-table-cell-content-with-inline-link bolt-table-cell-content flex-row flex-center"
          >
            <div
              class="flex-row flex-start"
            >
              <svg
                aria-labelledby="__bolt-status-1-desc"
                class="bolt-status flex-noshrink failed icon-large-margin status-icon"
                height="16"
                role="img"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <desc
                  id="__bolt-status-1-desc"
                >
                  Failed
                </desc>
                <circle
                  cx="8"
                  cy="8"
                  r="8"
                />
                <path
                  d="M10.984 5.004a.9.9 0 0 1 0 1.272L9.27 7.99l1.74 1.741a.9.9 0 1 1-1.272 1.273l-1.74-1.741-1.742 1.74a.9.9 0 1 1-1.272-1.272l1.74-1.74-1.713-1.714a.9.9 0 0 1 1.273-1.273l1.713 1.713 1.714-1.713a.9.9 0 0 1 1.273 0z"
                  fill="#fff"
                />
              </svg>
              <div
                class="flex-column wrap-text"
              >
                <a
                  class="bolt-table-inline-link bolt-table-link no-underline-link bolt-link"
                  href="https://dev.azure.com/org/project/_build/results?buildId=1"
                  rel="noopener"
                  tabindex="0"
                  target="_top"
                >
                  Build 123
                </a>
                <div
                  class="finish-date"
                >
                  <span
                    aria-label="Jun 29 at 12:00 PM Jun 29, 2025 at 12:00 PM UTC"
                    class="text-ellipsis"
                  >
                    <time
                      class="bolt-time-item white-space-nowrap"
                      datetime="2025-06-29T12:00:00.000Z"
                    >
                      Jun 29 at 12:00 PM
                    </time>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
